generator/validate/gen.go

```diff
@@ -0,0 +1,178 @@
+		} else {
+			unionMembers = append(unionMembers, unionField.Name)
+		}
+	}
+	if unionDiscriminatorField == nil {
+		root.AddError(loader.ErrFromNode(fmt.Errorf(
+			"Union `%v` should have 1 union discriminator. See here for details: https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190325-unions.md#proposal",
+			typeToCheck.Name), typeToCheck.RawSpec))
+		return
+	}
+
+	if unionDiscriminatorField.Markers.Get("optional") == nil {
+		root.AddError(loader.ErrFromNode(fmt.Errorf(
+			"In union `%v` the union discriminator `%v` should have the `+optional` comment marker",
+			typeToCheck.Name,
+			unionDiscriminatorField.Name), unionDiscriminatorField.RawField))
+	}
+	if !strings.Contains(unionDiscriminatorField.Tag.Get("json"), ",omitempty") {
+		root.AddError(loader.ErrFromNode(fmt.Errorf(
+			"In union `%v` the union discriminator `%v` should contain the `omitempty` option in its `json` tag, since it is expected to be an optional field",
+			typeToCheck.Name,
+			unionDiscriminatorField.Name), unionDiscriminatorField.RawField))
+	}
+
+	wrongTypeError := loader.ErrFromNode(fmt.Errorf(
+		"In union `%v` the union discriminator `%v` should have a `string` type, or a type (defined in the same package) whose underlying type is a string",
+		typeToCheck.Name,
+		unionDiscriminatorField.Name), unionDiscriminatorField.RawField)
+
+	discriminatorTypeRef, typeFound := root.TypesInfo.Types[unionDiscriminatorField.RawField.Type]
+	if !typeFound {
+		root.AddError(wrongTypeError)
+		return
+	}
+
+	underlyingType, underlyingTypeIsBasic := discriminatorTypeRef.Type.Underlying().(*types.Basic)
+	if !underlyingTypeIsBasic || underlyingType.Kind() != types.String {
+		root.AddError(wrongTypeError)
+		return
+	}
+
+	getEnumMarker := func(markers markers.MarkerValues) *crdmarkers.Enum {
+		enumMarkerIf := markers.Get("kubebuilder:validation:Enum")
+		if enumMarkerIf == nil {
+			return nil
+		}
+		enumMarker := enumMarkerIf.(crdmarkers.Enum)
+		return &enumMarker
+	}
+
+	var fieldEnumMarkerPtr *crdmarkers.Enum = getEnumMarker(unionDiscriminatorField.Markers)
+	var typeEnumMarkerPtr *crdmarkers.Enum = nil
+	var enumTypeInfo *markers.TypeInfo = nil
+
+	if enumTypeNamed, isNamed := discriminatorTypeRef.Type.(*types.Named); isNamed {
+		enumTypeInfo = packageTypes[enumTypeNamed.Obj().Name()]
+		if enumTypeInfo != nil {
+			typeEnumMarkerPtr = getEnumMarker(enumTypeInfo.Markers)
+		} else {
+			root.AddError(loader.ErrFromNode(fmt.Errorf(
+				"In union `%v` the union discriminator `%v` should have a string-based type defined in the same package",
+				typeToCheck.Name,
+				unionDiscriminatorField.Name), unionDiscriminatorField.RawField))
+		}
+	}
+
+	if fieldEnumMarkerPtr != nil && typeEnumMarkerPtr != nil {
+		root.AddError(loader.ErrFromNode(fmt.Errorf(
+			"Type `%v` should not define a `kubebuilder:validation:Enum` annotation, since it is already defined in the discriminator `%v` of union `%v`",
+			enumTypeInfo.Name,
```

Comment:
enumTypeInfo is nullable here. I assume return is missing after line 141

Reply:
Well, not really. In this case we know that `typeEnumMarkerPtr != nil`, and the case when `typeEnumMarkerPtr` can be set to a non-nil pointer, is precisely when `enumTypeInfo != nil`.